components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java

```diff
@@ -1826,97 +1821,67 @@ public int compare(Tag o1, Tag o2) {
-        		//a registry exception is thrown when the tenant store is accessed in anonymous mode.
-        		//This fix checks whether query resource available in the registry. If not
-        		// give a warn.
-				if (userRegistry != null && !userRegistry.resourceExists(tagsQueryPath)) {
-					log.warn("Failed to retrieve tags query resource at " + tagsQueryPath);
-					return tagSet == null ? Collections.EMPTY_SET : tagSet;
-				}
-			} catch (RegistryException e1) {
-                // Even if we should ignore this exception, we are logging this as a warn log.
-                // The reason is that, this error happens when we try to add some additional logs in an error
-                // scenario and it does not affect the execution path.
-                log.warn("Unable to execute the resource exist method for tags query resource path : " + tagsQueryPath,
-                         e1);
-            }
-            handleException("Failed to get all the tags", e);
-        } catch (UserStoreException e) {
-            handleException("Failed to get all the tags", e);
+        } catch (APIPersistenceException e) {
+            String msg = "Failed to get API tags";
+            throw new APIManagementException(msg, e);
         }
-
         return tagSet;
     }
 
+    @Override
+    public List<APICategory> getAllCategories(String organizationId) throws APIManagementException {
+        Organization org = new Organization(organizationId);
+        String userName = (userNameWithoutChange != null)? userNameWithoutChange: username;
+        String[] roles = APIUtil.getListOfRoles(userName);
+        Map<String, Object> properties = APIUtil.getUserProperties(userName);
+        UserContext userCtx = new UserContext(userNameWithoutChange, org, properties, roles);
+        List<APICategory> categoriesList = new ArrayList<>();
+        try {
+            DevPortalAPISearchResult searchAPIs = apiPersistenceInstance.searchAPIsForDevPortal(org, "",
+                    0, 25, userCtx);
+            SortedSet<Object> apiSet = new TreeSet<>(new APIAPIProductNameComparator());
+            if (searchAPIs != null) {
+                List<DevPortalAPIInfo> list = searchAPIs.getDevPortalAPIInfoList();
+                List<Object> apiList = new ArrayList<>();
+                for (DevPortalAPIInfo devPortalAPIInfo : list) {
+                    API mappedAPI = APIMapper.INSTANCE.toApi(devPortalAPIInfo);
+                    List<APICategory> mappedAPIApiCategories = mappedAPI.getApiCategories();
+                    for (APICategory apiCategory: mappedAPIApiCategories) {
+                        categoriesList.add(apiCategory);
+                    }
+                }
+            }
+            return  categoriesList;
+        } catch (APIPersistenceException e) {
+            String msg = "Failed to get API tags";
+            throw new APIManagementException(msg, e);
+        }
```

Comment:
Catagories are stored in AM_DB with AM_API_CATEGORIES under each TenantID in regular APIM. I think we'll need to utilize this table here as well with some changes rather than traversing all the APIs finding each category. But this might require some changes in these tables. It is ok to have this implementation temporary but we must do a rethinking proper implementation after APIM 4.0 is released and reduce these deviation of tables. What we can do here is, in regular APIM also, in the table we can have ORGANIZATION_ID instead of TENANT_ID. An have an abstraction to use tenantID or organizationID in two different implementations.
Please log an issue to track this.
@pubudu538 FYI

Reply:
The approach @malinthaprasan suggested is the best way to tackle this.  Here we are getting the categories from the registry(like other flows) without any meta information. When we are changing with the DB approach lots of changes need to be incorporated starting from Admin rest APIs to DAO layer changes. As you mentioned let's initiate this as a separate effort.

Git issue : https://github.com/wso2/product-apim/issues/10488